### PR TITLE
Add tests for various issues

### LINKED ISSUE #11740: Unexplained lifetime error in closure type
In the following test, removing the type annotation from `attr` produces lifetime errors:

```
#[feature(managed_boxes)];

use std::cast;
use std::unstable::raw::Box;

struct Attr {
    name: ~str,
    value: ~str,
}

struct Element {
    attrs: ~[@mut Attr],
}

impl Element {
    pub unsafe fn get_attr<'a>(&'a self, name: &str) {
        self.attrs.iter().find(|attr| { // To remove error: |attr: & & 'a @mut Attr| {
                let attr: ***Box<Attr> = cast::transmute(attr);
                true
        });
    }
}

pub fn main() {
    let element = Element {
        attrs: ~[],
    };
    let _ = unsafe { element.get_attr("foo") };
}
```

Error message:

```
Running /home/nmatsakis/versioned/rust-1/build/i686-unknown-linux-gnu/stage2/bin/rustc:
/home/nmatsakis/tmp/foo.rs:17:9: 17:27 error: cannot infer an appropriate lifetime for autoref due to conflicting requirements
/home/nmatsakis/tmp/foo.rs:17         self.attrs.iter().find(|attr| { // |attr: & & 'a @Attr| {
                                      ^~~~~~~~~~~~~~~~~~
/home/nmatsakis/tmp/foo.rs:17:9: 17:19 note: first, the lifetime cannot outlive the expression at 17:8...
/home/nmatsakis/tmp/foo.rs:17         self.attrs.iter().find(|attr| { // |attr: & & 'a @Attr| {
                                      ^~~~~~~~~~
/home/nmatsakis/tmp/foo.rs:17:9: 17:19 note: ...so that automatically reference is valid at the time of borrow
/home/nmatsakis/tmp/foo.rs:17         self.attrs.iter().find(|attr| { // |attr: & & 'a @Attr| {
                                      ^~~~~~~~~~
/home/nmatsakis/tmp/foo.rs:17:9: 17:27 note: but, the lifetime must be valid for the method call at 17:8...
/home/nmatsakis/tmp/foo.rs:17         self.attrs.iter().find(|attr| { // |attr: & & 'a @Attr| {
                                      ^~~~~~~~~~~~~~~~~~
/home/nmatsakis/tmp/foo.rs:17:9: 17:19 note: ...so that method receiver is valid for the method call
/home/nmatsakis/tmp/foo.rs:17         self.attrs.iter().find(|attr| { // |attr: & & 'a @Attr| {
                                      ^~~~~~~~~~
error: aborting due to previous error
task 'rustc' failed at 'explicit failure', /home/nmatsakis/versioned/rust-1/src/libsyntax/diagnostic.rs:75
task '<main>' failed at 'explicit failure', /home/nmatsakis/versioned/rust-1/src/librustc/lib.rs:448
```

### LINKED ISSUE #19601: "error: illegal recursive type"
Is the restriction on recursive types really necessary?

It makes sense with _structs_, because recursive structs would mean infinite-sized structs. But with _types_, it seems unnecessary. After a brief discussion on the Rust IRC, someone told me that this is probably a bug.
## Reduced Test Case

``` rust
trait A<T> {}

trait DoesAforB<T> {}

// no way to express
// struct B<T> where B<T>: A<B<T>>;

struct B<T: DoesAforB<B<T>>>;

fn main() {}
```
## Output

```
rec.rs:8:23: 8:27 error: illegal recursive type; insert an enum or struct in the cycle, if this is desired
rec.rs:8 struct B<T: DoesAforB<B<T>>>;
                               ^~~~
```
## rustc version

`rustc 0.13.0-nightly (336349c93 2014-11-17 20:37:19 +0000)`

### LINKED ISSUE #22603: ICE implementing Fn trait for unit struct
The following programs generate the ICE with a slightly different message. May be related to #22565 and #22566 

``` rust
#![feature(unboxed_closures)]

pub struct Foo;

impl<A> Fn<(A,)> for Foo {
    type Output = ();
    extern "rust-call" fn call(&self, (_,): (A,)) {
    }
}

pub fn main() {
    println!("{:?}", Foo("bar"));
}
```

```
error: internal compiler error: unexpected panic
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: http://doc.rust-lang.org/complement-bugreport.html
note: run with `RUST_BACKTRACE=1` for a backtrace
thread 'rustc' panicked at 'assertion failed: `(left == right) && (right == left)` (left: `1`, right: `0`)', /Users/dwm/dev/freebroccolo/rust/src/librustc_trans/trans/adt.rs:1120

stack backtrace:
   1:        0x10f16fe23 - sys::backtrace::write::h3d6ead7c1184a95aYoB
   2:        0x10f19db85 - panicking::on_panic::h4ff071a70fb2a50dAGK
   3:        0x10f0cf188 - rt::unwind::begin_unwind_inner::h9fcba14039c4ae30QnK
   4:        0x10f0cf99e - rt::unwind::begin_unwind_fmt::h8a64dc3276ab80e6mmK
   5:        0x10b78ef5a - trans::adt::build_const_struct::h9f178ea5cb15adefd8J
   6:        0x10b6feb12 - trans::adt::trans_const::he12f11c052fd51c6qXJ
   7:        0x10b6fa1cb - trans::consts::const_expr_unadjusted::h6f4b4da94e54f806amo
   8:        0x10b6fca28 - trans::consts::const_expr::he362f3efa474e33578n
   9:        0x10b6fa146 - trans::consts::const_expr_unadjusted::h6f4b4da94e54f806amo
  10:        0x10b6fca28 - trans::consts::const_expr::he362f3efa474e33578n
  11:        0x10b6fe213 - iter::IteratorExt::fold::h16846351534471258687
  12:        0x10b6f893f - trans::consts::const_expr_unadjusted::h6f4b4da94e54f806amo
  13:        0x10b6fca28 - trans::consts::const_expr::he362f3efa474e33578n
  14:        0x10b68ed84 - trans::consts::get_const_expr_as_global::h0122b2bc5a07a52ed6n
  15:        0x10b643a17 - trans::expr::trans::h181c848f1f91f5c7BMh
  16:        0x10b734ee4 - trans::_match::trans_match_inner::h6e02f6c0e6a4b098zMx
  17:        0x10b691ee0 - trans::expr::trans_rvalue_dps_unadjusted::h5f563e68c9cce0602Wi
  18:        0x10b690362 - trans::expr::trans_unadjusted::h1caf7e41a27a61e7mqi
  19:        0x10b643ed0 - trans::expr::trans::h181c848f1f91f5c7BMh
  20:        0x10b6a71e3 - trans::expr::trans_addr_of::h3700c26ce80cf333tNj
  21:        0x10b6908f7 - trans::expr::trans_unadjusted::h1caf7e41a27a61e7mqi
  22:        0x10b643ed0 - trans::expr::trans::h181c848f1f91f5c7BMh
  23:        0x10b680aa5 - trans::callee::trans_args::h670e438b49426a11rjh
  24:        0x10b687394 - trans::callee::trans_call_inner::h12819191760544836083
  25:        0x10b692aed - trans::expr::trans_rvalue_dps_unadjusted::h5f563e68c9cce0602Wi
  26:        0x10b690362 - trans::expr::trans_unadjusted::h1caf7e41a27a61e7mqi
  27:        0x10b643ed0 - trans::expr::trans::h181c848f1f91f5c7BMh
  28:        0x10b680aa5 - trans::callee::trans_args::h670e438b49426a11rjh
  29:        0x10b687394 - trans::callee::trans_call_inner::h12819191760544836083
  30:        0x10b692aed - trans::expr::trans_rvalue_dps_unadjusted::h5f563e68c9cce0602Wi
  31:        0x10b6426eb - trans::expr::trans_into::h3a3c8b3d53f8afd1jGh
  32:        0x10b64153b - trans::controlflow::trans_stmt_semi::h26c41836c06fdbd36de
  33:        0x10b6431a3 - trans::controlflow::trans_block::hd7ca24723db0b5fbXee
  34:        0x10b7191f9 - trans::base::trans_closure::hbc388ffaaf4af161Fiu
  35:        0x10b62d4d6 - trans::base::trans_fn::hd5c4904336306603Wtu
  36:        0x10b62883d - trans::base::trans_item::h6fa4471c5f48a0b4PSu
  37:        0x10b71fbdc - trans::base::trans_crate::h15ffd65af4da49d1fQv
  38:        0x10b4ac638 - driver::phase_4_translate_to_llvm::h19f970ffecf43574iPa
  39:        0x10b4864f6 - driver::compile_input::hd5773fb44ef48d33Eba
  40:        0x10b560414 - run_compiler::h70410abf4d17bf43Rbc
  41:        0x10b55d495 - thunk::F.Invoke<A, R>::invoke::h1011432548731059940
  42:        0x10b55c130 - rt::unwind::try::try_fn::h9217449077233910915
  43:        0x10f21a289 - rust_try_inner
  44:        0x10f21a276 - rust_try
  45:        0x10b55c829 - thunk::F.Invoke<A, R>::invoke::h35832703021491210
  46:        0x10f186373 - sys::thread::thread_start::h5036bdc0fe436b19CWF
  47:     0x7fff82a3f268 - _pthread_body
  48:     0x7fff82a3f1e5 - _pthread_body
```

``` rust
#![feature(unboxed_closures)]

pub struct Foo;

impl<A> Fn<(A,)> for Foo {
    type Output = A;
    extern "rust-call" fn call(&self, (arg,): (A,)) -> A {
        arg
    }
}

pub fn main() {
    println!("{:?}", Foo("bar"));
}
```

```
error: internal compiler error: adt::represent_type called on non-ADT type: &'static str
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: http://doc.rust-lang.org/complement-bugreport.html
note: run with `RUST_BACKTRACE=1` for a backtrace
thread 'rustc' panicked at 'Box<Any>', /Users/dwm/dev/freebroccolo/rust/src/libsyntax/diagnostic.rs:189

stack backtrace:
   1:        0x107908e23 - sys::backtrace::write::h3d6ead7c1184a95aYoB
   2:        0x107936b85 - panicking::on_panic::h4ff071a70fb2a50dAGK
   3:        0x107868188 - rt::unwind::begin_unwind_inner::h9fcba14039c4ae30QnK
   4:        0x106f32b2f - rt::unwind::begin_unwind::h6422578009397946341
   5:        0x106f3345d - diagnostic::Handler::bug::he3fe83188ec467dceeF
   6:        0x1047a7568 - session::Session::bug::hc6ca12d54ef8a384z3s
   7:        0x103f16223 - trans::adt::represent_type_uncached::h3807220f4faf82436mI
   8:        0x103de3790 - trans::adt::represent_type::hdcbcf49d3304f11cGjI
   9:        0x103e871a0 - trans::consts::const_expr_unadjusted::h6f4b4da94e54f806amo
  10:        0x103e89a28 - trans::consts::const_expr::he362f3efa474e33578n
  11:        0x103e87146 - trans::consts::const_expr_unadjusted::h6f4b4da94e54f806amo
  12:        0x103e89a28 - trans::consts::const_expr::he362f3efa474e33578n
  13:        0x103e8b213 - iter::IteratorExt::fold::h16846351534471258687
  14:        0x103e8593f - trans::consts::const_expr_unadjusted::h6f4b4da94e54f806amo
  15:        0x103e89a28 - trans::consts::const_expr::he362f3efa474e33578n
  16:        0x103e1bd84 - trans::consts::get_const_expr_as_global::h0122b2bc5a07a52ed6n
  17:        0x103dd0a17 - trans::expr::trans::h181c848f1f91f5c7BMh
  18:        0x103ec1ee4 - trans::_match::trans_match_inner::h6e02f6c0e6a4b098zMx
  19:        0x103e1eee0 - trans::expr::trans_rvalue_dps_unadjusted::h5f563e68c9cce0602Wi
  20:        0x103e1d362 - trans::expr::trans_unadjusted::h1caf7e41a27a61e7mqi
  21:        0x103dd0ed0 - trans::expr::trans::h181c848f1f91f5c7BMh
  22:        0x103e341e3 - trans::expr::trans_addr_of::h3700c26ce80cf333tNj
  23:        0x103e1d8f7 - trans::expr::trans_unadjusted::h1caf7e41a27a61e7mqi
  24:        0x103dd0ed0 - trans::expr::trans::h181c848f1f91f5c7BMh
  25:        0x103e0daa5 - trans::callee::trans_args::h670e438b49426a11rjh
  26:        0x103e14394 - trans::callee::trans_call_inner::h12819191760544836083
  27:        0x103e1faed - trans::expr::trans_rvalue_dps_unadjusted::h5f563e68c9cce0602Wi
  28:        0x103e1d362 - trans::expr::trans_unadjusted::h1caf7e41a27a61e7mqi
  29:        0x103dd0ed0 - trans::expr::trans::h181c848f1f91f5c7BMh
  30:        0x103e0daa5 - trans::callee::trans_args::h670e438b49426a11rjh
  31:        0x103e14394 - trans::callee::trans_call_inner::h12819191760544836083
  32:        0x103e1faed - trans::expr::trans_rvalue_dps_unadjusted::h5f563e68c9cce0602Wi
  33:        0x103dcf6eb - trans::expr::trans_into::h3a3c8b3d53f8afd1jGh
  34:        0x103dce53b - trans::controlflow::trans_stmt_semi::h26c41836c06fdbd36de
  35:        0x103dd01a3 - trans::controlflow::trans_block::hd7ca24723db0b5fbXee
  36:        0x103ea61f9 - trans::base::trans_closure::hbc388ffaaf4af161Fiu
  37:        0x103dba4d6 - trans::base::trans_fn::hd5c4904336306603Wtu
  38:        0x103db583d - trans::base::trans_item::h6fa4471c5f48a0b4PSu
  39:        0x103eacbdc - trans::base::trans_crate::h15ffd65af4da49d1fQv
  40:        0x103c3a638 - driver::phase_4_translate_to_llvm::h19f970ffecf43574iPa
  41:        0x103c144f6 - driver::compile_input::hd5773fb44ef48d33Eba
  42:        0x103cee414 - run_compiler::h70410abf4d17bf43Rbc
  43:        0x103ceb495 - thunk::F.Invoke<A, R>::invoke::h1011432548731059940
  44:        0x103cea130 - rt::unwind::try::try_fn::h9217449077233910915
  45:        0x1079b3289 - rust_try_inner
  46:        0x1079b3276 - rust_try
  47:        0x103cea829 - thunk::F.Invoke<A, R>::invoke::h35832703021491210
  48:        0x10791f373 - sys::thread::thread_start::h5036bdc0fe436b19CWF
  49:     0x7fff82a3f268 - _pthread_body
  50:     0x7fff82a3f1e5 - _pthread_body
```

### LINKED ISSUE #22789: ICE static call to invalid vtable
I was trying to compile the following code:

``` rust
#![feature(core)]
#![feature(unboxed_closures)]
enum Thunk<A, F: Fn<A>> { Unevaluated(F, A), Evaluated(F::Output), }
impl <A, F: Fn<A>> Thunk<A, F> {
    fn eval(self) -> (Self, F::Output) where F::Output: Copy {
        match self {
            Thunk::Unevaluated(f, a) => {
                let r = Fn::call(&f, a);
                (Thunk::Evaluated(r), r)
            }
            Thunk::Evaluated(r) => (self, r),
        }
    }
}
fn main() {
    let (_, v) = Thunk::Unevaluated(|x: i32| { x + 1 }, (3,)).eval();
    println!("{}" , v);
}
```

And got the following compiler error:

```
error: internal compiler error: static call to invalid vtable: VtableClosure(DefId { krate: 0, node: 97 },Substs[types=[[];[];[]], regions=eras
ed])
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
note: run with `RUST_BACKTRACE=1` for a backtrace
thread 'rustc' panicked at 'Box<Any>', /home/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libsyntax/diagnostic.rs:189

stack backtrace:
   1:     0x7f77423cc210 - sys::backtrace::write::h58f3688f47ea9825KlC
   2:     0x7f77423f4ac0 - panicking::on_panic::h6a53d3220988f2eebYL
   3:     0x7f7742335720 - rt::unwind::begin_unwind_inner::h72a0c6e103bfcdc3zCL
   4:     0x7f773f5ffde0 - rt::unwind::begin_unwind::h2955836616559227712
   5:     0x7f773f6005a0 - diagnostic::Handler::bug::h3ef6141b1fad14d7r5E
   6:     0x7f774034e2c0 - session::Session::bug::h912125adfec44a01Snt
   7:     0x7f774103af70 - trans::meth::trans_static_method_callee::hee20b523e376bdd9mjz
   8:     0x7f7741037e40 - trans::callee::trans::h6f90cc6896233847Kgg
   9:     0x7f774104e040 - trans::callee::trans_call_inner::h364678376362576570
  10:     0x7f7741057990 - trans::expr::trans_rvalue_dps_unadjusted::h706f4998bd14a2629Wi
  11:     0x7f774100ccc0 - trans::expr::trans_into::hdab96306e19b21bcpGh
  12:     0x7f7741111450 - trans::_match::mk_binding_alloca::h10439043131118879359
  13:     0x7f774100c400 - trans::base::init_local::hcdcc29de8e78259fwst
  14:     0x7f774100d830 - trans::controlflow::trans_block::h5bbb67526ff90a125ee
  15:     0x7f7741057990 - trans::expr::trans_rvalue_dps_unadjusted::h706f4998bd14a2629Wi
  16:     0x7f774100ccc0 - trans::expr::trans_into::hdab96306e19b21bcpGh
  17:     0x7f77410f9260 - trans::_match::trans_match_inner::h6651c4e019791e16aMx
  18:     0x7f7741057990 - trans::expr::trans_rvalue_dps_unadjusted::h706f4998bd14a2629Wi
  19:     0x7f774100ccc0 - trans::expr::trans_into::hdab96306e19b21bcpGh
  20:     0x7f774100d830 - trans::controlflow::trans_block::h5bbb67526ff90a125ee
  21:     0x7f77410dad70 - trans::base::trans_closure::h8de4dc5484c0f3bephu
  22:     0x7f7740ff9b30 - trans::base::trans_fn::h734239fd8d4c689fGsu
  23:     0x7f7740ffa090 - trans::monomorphize::monomorphic_fn::he0cb46049f2b73776Ad
  24:     0x7f774103c170 - trans::callee::trans_fn_ref_with_substs::h5a9f25c47377da3bYDg
  25:     0x7f774103abf0 - trans::callee::trans_fn_ref::h49bed2d85d0bd0955rg
  26:     0x7f7741050b90 - trans::meth::trans_method_callee::had7e320247e6010e3dz
  27:     0x7f774104f640 - trans::callee::trans_call_inner::h11271674016489502486
  28:     0x7f7741057990 - trans::expr::trans_rvalue_dps_unadjusted::h706f4998bd14a2629Wi
  29:     0x7f7741056dd0 - trans::expr::trans_unadjusted::h4aa6c42f21ab6029sqi
  30:     0x7f774100e360 - trans::expr::trans::h522d304fced36cc2HMh
  31:     0x7f774100c400 - trans::base::init_local::hcdcc29de8e78259fwst
  32:     0x7f774100d830 - trans::controlflow::trans_block::h5bbb67526ff90a125ee
  33:     0x7f77410dad70 - trans::base::trans_closure::h8de4dc5484c0f3bephu
  34:     0x7f7740ff9b30 - trans::base::trans_fn::h734239fd8d4c689fGsu
  35:     0x7f7740ff4d50 - trans::base::trans_item::h653e553ae00342d4zRu
  36:     0x7f77410e22e0 - trans::base::trans_crate::hffa644676f991a591Ov
  37:     0x7f7742a534f0 - driver::phase_4_translate_to_llvm::h7bbc12961d2d7c58qPa
  38:     0x7f7742a2d2d0 - driver::compile_input::hfb1bb38c59866d0fGba
  39:     0x7f7742afc510 - run_compiler::hf2a7d9848e51e3b5Zbc
  40:     0x7f7742afae10 - thunk::F.Invoke<A, R>::invoke::h12698859262113329290
  41:     0x7f7742af9d00 - rt::unwind::try::try_fn::h10854600533917764358
  42:     0x7f7742461880 - rust_try_inner
  43:     0x7f7742461870 - rust_try
  44:     0x7f7742afa000 - thunk::F.Invoke<A, R>::invoke::h4174228679526982116
  45:     0x7f77423e0b60 - sys::thread::thread_start::h60ae7cb39d135ef7R4G
  46:     0x7f773c2be0c0 - start_thread
  47:     0x7f7741fadfd9 - __clone
  48:                0x0 - <unknown>
```

Version info:

```
$ rustc --version --verbose
rustc 1.0.0-nightly (2b01a37ec 2015-02-21) (built 2015-02-22)
binary: rustc
commit-hash: 2b01a37ec38db9301239f0c0abcf3c695055b0ff
commit-date: 2015-02-21
build-date: 2015-02-22
host: x86_64-unknown-linux-gnu
release: 1.0.0-nightly
```

### LINKED ISSUE #26614: Associated type in constant not normalized correctly
``` rust
trait Mirror { type It; }
impl<T> Mirror for T { type It = Self; }
fn main() {
    let c : <u32 as Mirror>::It = 5;
    const CCCC : <u32 as Mirror>::It = 5;
}
```

```
<anon>:5:40: 5:41 error: mismatched types:
 expected `<u32 as Mirror>::It`,
    found `_`
(expected associated type,
    found integral variable) [E0308]
<anon>:5     const CCCC : <u32 as Mirror>::It = 5;
                                                ^
<anon>:5:40: 5:41 help: see the detailed explanation for E0308
error: aborting due to previous error
```

### PR DESCRIPTION
**Pull request by MaloJaffre**
Fixes #11740.
Fixes #19601.
Fixes #22603
Fixes #22789.
Fixes #26614.

r? @Mark-Simulacrum.

### CODE REVIEW

File: src/test/compile-fail/issue-11740.rs [hunk 1]
```diff
@@ -0,0 +1,38 @@
+// Copyright 2014 The Rust Project Developers. See the COPYRIGHT
```
Comment (Mark-Simulacrum) [inline src/test/compile-fail/issue-11740.rs]: Could you fix up the copyright dates to 2017? I don't know if it matters, but it seems wrong to back-date copyrights.

**Review by Mark-Simulacrum — CHANGES_REQUESTED** [review]

**Comment by Mark-Simulacrum** [comment]
r=me aside from the copyright dates, thanks!

@bors rollup

**Comment by MaloJaffre** [comment]
Fixed, thanks for review @Mark-Simulacrum!

**Review by Mark-Simulacrum — APPROVED** [review]

**Pull request merged**